streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -304,162 +392,79 @@ public void flush() {
-        return partition == null ? taskId.partition : partition.partition();
-    }
-
-    void registerGlobalStateStores(final List<StateStore> stateStores) {
-        log.debug("Register global stores {}", stateStores);
-        for (final StateStore stateStore : stateStores) {
-            globalStores.put(stateStore.name(), Optional.of(stateStore));
-        }
-    }
 
-    @Override
-    public StateStore getGlobalStore(final String name) {
-        return globalStores.getOrDefault(name, Optional.empty()).orElse(null);
-    }
-
-    Collection<TopicPartition> changelogPartitions() {
-        return unmodifiableList(changelogPartitions);
-    }
-
-    void ensureStoresRegistered() {
-        for (final Map.Entry<String, Optional<StateStore>> entry : registeredStores.entrySet()) {
-            if (!entry.getValue().isPresent()) {
-                throw new IllegalStateException(
-                    "store [" + entry.getKey() + "] has not been correctly registered. This is a bug in Kafka Streams."
-                );
+        final Map<TopicPartition, Long> checkpointingOffsets = new HashMap<>();
+        for (final StateStoreMetadata storeMetadata : stores.values()) {
+            // store is logged, persistent, and has a valid current offset
+            if (storeMetadata.changelogPartition != null &&
+                storeMetadata.stateStore.persistent() &&
+                storeMetadata.offset != null) {
+                checkpointingOffsets.put(storeMetadata.changelogPartition, storeMetadata.offset);
             }
         }
-    }
-
-    private Set<TopicPartition> validCheckpointableTopics() {
-        // it's only valid to record checkpoints for registered stores that are both persistent and change-logged
-
-        final Set<TopicPartition> result = new HashSet<>(storeToChangelogTopic.size());
-        for (final Map.Entry<String, String> storeToChangelog : storeToChangelogTopic.entrySet()) {
-            final String storeName = storeToChangelog.getKey();
-            if (registeredStores.containsKey(storeName)
-                && registeredStores.get(storeName).isPresent()
-                && registeredStores.get(storeName).get().persistent()) {
 
-                final String changelogTopic = storeToChangelog.getValue();
-                result.add(new TopicPartition(changelogTopic, getPartition(changelogTopic)));
-            }
+        log.debug("Writing checkpoint: {}", checkpointingOffsets);
+        try {
+            checkpointFile.write(checkpointingOffsets);
+        } catch (final IOException e) {
+            log.warn("Failed to write offset checkpoint file to [{}]", checkpointFile, e);
         }
-        return result;
     }
 
-    private static Map<TopicPartition, Long> validCheckpointableOffsets(
-        final Map<TopicPartition, Long> checkpointableOffsets,
-        final Set<TopicPartition> validCheckpointableTopics) {
+    private StateStoreMetadata findStore(final TopicPartition changelogPartition) {
```

Comment:
maybe it doesn't matter, but this method makes the only usage an n^2 algorithm. If we instead inverted the `stores` collection and used it for lookups in `register`, it would be o(n)

Reply:
I've thought about it: inverting the `stores` collection makes other calls that depends on the store name more complicated, while keeping two collections indexed by storeName / changelog partition is not much worthy since within a task there are usually no more than 10 stores so this n^2 algorithm should not be a big deal.